plugin/restore/mysql/mysql.go

```diff
@@ -457,87 +484,144 @@ func getPITROldDatabaseName(database string, suffixTs int64) string {
+		binlogFilesLocalValid = append(binlogFilesLocalValid, file)
+	}
+	binlogFilesLocalValidSorted, err := sortBinlogFiles(binlogFilesLocalValid)
+	if err != nil {
+		return err
+	}
+
+	// Find if some local binlog file's first binlog event ts is already larger than targetTs.
+	// If so, we do not need to download the following binlog files on server.
+	var foundLocalFileLTTargetTs bool
+	var localFileExceedsTargetTs BinlogFile
+	for _, file := range binlogFilesLocalValidSorted {
+		firstBinlogEvent := api.BinlogInfo{FileName: file.Name, Position: 4}
+		eventTs, err := r.parseLocalBinlogEventTimestamp(ctx, firstBinlogEvent)
+		path := filepath.Join(r.binlogDir, file.Name)
+		if err != nil {
+			log.Error("Failed to parse the first binlog event timestamp for local binlog file",
+				zap.String("path", path),
+				zap.Error(err))
+			return fmt.Errorf("failed to parse the first binlog event timestamp for local binlog file[%s], error[%w]", path, err)
 		}
-		if err := r.downloadBinlogFile(ctx, serverFile, false); err != nil {
-			return fmt.Errorf("cannot sync binlog %s, error: %w", serverFile.Name, err)
+		if eventTs > targetTs {
+			log.Info("Found a local binlog file whose first event's timestamp is larger than targetTs",
+				zap.Any("binlogFile", file),
+				zap.Int64("targetTs", targetTs),
+				zap.Int64("eventTs", eventTs))
+			foundLocalFileLTTargetTs = true
+			localFileExceedsTargetTs = file
+			break
 		}
 	}
 
-	return nil
-}
+	log.Debug("Download binlog file list", zap.Array("list", ZapBinlogFiles(downloadFileList)))
+	for _, file := range downloadFileList {
+		log.Debug("Downloading file", zap.Any("binlogFile", file))
+		if foundLocalFileLTTargetTs && localFileExceedsTargetTs.LE(file) {
+			log.Debug("Skip downloading file, because there's already a local binlog file whose first eventTs > targetTs", zap.String("binlogFile", file.Name))
+			break
+		}
 
-// IncrementalFetchAllBinlogFiles fetches all the binlog files that on server but not on local.
-// TODO(zp): It is not yet supported to synchronize some binlog files.
-// The current practice is to download all binlog files, but in the future we hope to only synchronize to the PITR time point
-func (r *Restore) IncrementalFetchAllBinlogFiles(ctx context.Context) error {
-	if err := r.FetchArchivedBinlogFiles(ctx); err != nil {
-		return err
-	}
-	latestBinlogFileOnServer, err := r.getLatestBinlogFileMeta(ctx)
-	if err != nil {
-		return err
+		isLast := file == latestBinlogFileOnServer
+		if err := r.downloadBinlogFile(ctx, file, isLast); err != nil {
+			return fmt.Errorf("failed to download binlog file[%s], error[%w]", file.Name, err)
+		}
+		// We parsed the first binlog event's timestamp of the newly downloaded binlog file.
+		// If the timestamp > targetTs, we skip downloading the following binlog files, because we have already
+		// downloaded all the binlog files needed before targetTs.
+		firstBinlogEvent := api.BinlogInfo{FileName: file.Name, Position: 4}
```

Comment:
This function is quite large and complex. Think of a way to do unit testing for some part of it.

Reply:
There are some integration test cases.
I'll extract the download list generation algorithm and write some UT for it.